docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,363 @@
+```
+
+##### Example response in JSON based protocol
+
+```
+git_promises 0.0.1 v1 json_based
+
+{"operation": "validate_promise", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}, "result": "valid"}
+
+log_info=Cloning 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'...
+log_info=Successfully cloned 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'
+{"operation": "evaluate_promise", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}, "result": "kept"}
+
+{"operation": "terminate", "result": "success"}
+```
+
+In the example above, log messages are sent to the agent on separate lines.
+The syntax is the same as for the line based protocol, described below.
+This is done so the agent can print the messages while the promise is evaluating.
+(In the example, cloning the repo takes a few seconds, and it is nice to get some feedback).
+
+Log messages formatted like this must be before the JSON message, and it is optional.
+You can also include log messages in the JSON data:
+
+```
+{
+  "operation": "evaluate_promise",
+  "promiser": "/opt/cfengine/masterfiles",
+  "attributes": {
+    "repo": "https://github.com/cfengine/masterfiles"}
+  },
+  "log": [
+    {
+      "level": "info",
+      "message": "Cloning 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'..."
+    }
+  ],
+  "result": "kept"
+}
+```
+
+(Formatted here for readability, JSON data must still be on a single line in protocol).
+
+#### Line based protocol
+
+The line based protocol supports a limited subset of features of CFEngine promises, which should be enough for most use cases.
+In particular, data passed to the module (through promise attributes) must be strings, without newlines.
+`slist`, `data` containers, strings with newlines are not supported.
+It is designed to be easy to implement in a language which doesn't have JSON support.
+
+A protocol message is a sequence of lines, separated by newline characters, and terminated by an extra newline character (empty line).
+Each line is a key-value pair, separated by `=`.
+Anything before the first `=` of a line is the key, anything after is the value.
+
+The key may not be empty, but the value may.
+The key must consist of lowercase letters and underscores, not numbers, spaces, quotes, or other symbols (`[a-z_]+`).
+The value may contain anything (including `=` signs) except for newlines and zero-bytes.
+
+##### Example requests in line based protocol
+
+```conf
+CFEngine 3.16.0 v1
+
+operation=validate_promise
+log_level=info
+promise_type=git
+promiser=/opt/cfengine/masterfiles
+attribute_repo=https://github.com/cfengine/masterfiles
+
+operation=evaluate_promise
+log_level=info
+promise_type=git
+promiser=/opt/cfengine/masterfiles
+attribute_repo=https://github.com/cfengine/masterfiles
+
+operation=terminate
+log_level=info
```

Comment:
what is the repetition of `log_level` in each block above for?

Reply:
This is a series of 4 requests. The `log_level` is a helper for the module to skip some log messages if they won't be printed anyway. The log level can change, cf-agent might request different log levels for different promises.